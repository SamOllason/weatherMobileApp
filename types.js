export const APIToken    = '03d6a8a8a5e2a10e41df0d808802c751';
export const error       = 'error';
export const defaultInfo = "--";
