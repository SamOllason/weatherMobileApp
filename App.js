import React, {Component} from 'react';
import ForecastList from './ForecastList';
import {StyleSheet, View, Alert} from 'react-native';
import CurrentWeatherPane from './CurrentWeatherPane';
import BackgroundImage from './BackgroundImage';
import BottomSection from './BottomSection';
import * as types from './types';

// const instructions = Platform.select({
//   ios: 'Press Cmd+R to reload,\n' + 'Cmd+D or shake for dev menu',
//   android:
//     'Double tap R on your keyboard to reload,\n' +
//     'Shake or press menu button for dev menu',
// });

const appId = '03d6a8a8a5e2a10e41df0d808802c751';

export default class App extends Component {

  constructor(props){
    super(props);

    this.state = {
      textInput: types.defaultInfo,
      city: types.defaultInfo,
      currentWeather: types.defaultInfo,
      temperature: types.defaultInfo,
      textColour: types.whitesmoke,

      forecast: [],
      modalVisible: false,
    };

    this.getWeather    = this.getWeather.bind(this);
    this.showModal     = this.showModal.bind(this);
    this.hideModal     = this.hideModal.bind(this);
    this._onTextChange = this._onTextChange.bind(this);
  }

  /**
   * Attempt to fetch current weather for using text input from user
   * If successful then set state and also fetch forecast for next 5 days.
   * If forecast request unsuccessful show alert with error message
   *
   * If current weather request not successful then show alert and do not make forecast request.
   */

  getWeather = async() => {

    const city = this.state.textInput;

    // use 'await' because function returns a promise
    const currentWeatherData = await this.getCurrentWeather(city);

    const currentWeatherRequestAllOk = this.checkDataStatusOk(currentWeatherData);

    if(!currentWeatherRequestAllOk) {
      // show alert and do not attempt to get forecast
      this.handleError(currentWeatherData);
    } else {
      this.updateCurrentWeather(currentWeatherData);

      // Since current weather fetch was a success, get forecast
      const forecastData      = await this.getForecast(city);
      const forecastDataAllOk = this.checkDataStatusOk(forecastData);

      if(!forecastDataAllOk){
        this.handleError(forecastData);
      } else {
        this.updateForecastDataInState(forecastData);
      }
    }
  };

  getCurrentWeather = async (city, token = types.APIToken) => {
    try {
      const currentWeatherResponse = await fetch(
          `https://api.openweathermap.org/data/2.5/weather?q=${city}&APPID=${token}&units=metric`
      );

      const currentWeatherResponseJson = await currentWeatherResponse.json();

      return currentWeatherResponseJson;

    } catch(err) {
      console.log(`current weather request Error: ${err}`);
      return err;
    }
  };

  checkDataStatusOk = (data) => {
    return data.cod === 200 || data.cod === "200";
  };

  updateCurrentWeather = (weatherData) => {

    // const currentWeather = weatherData.weather[0].main;

    const currentWeather = 'snow';

    const textColour  = this.getTextColour(currentWeather);

    this.setState({
      // Make it clear to user what city data is for.
      city: weatherData.name,
      temperature: weatherData.main.temp,
      currentWeather: currentWeather,
      textColour: textColour
    });
  };

  updateForecastDataInState = (forecastData) => {
    this.setState({
      forecast: forecastData.list,
    });
  };

  handleError = (data) => {
    if(data.message) {
      Alert.alert(data.message);
    } else {
      Alert.alert("Unknown error");
    }
  };

  getForecast = async (city) => {
    try {
      const forecastResponse = await fetch(
          `https://api.openweathermap.org/data/2.5/forecast?q=${city}&APPID=${appId}&units=metric`
      );
      const forecastResponseJson = await forecastResponse.json();

      return forecastResponseJson;

    } catch(err){
        console.log(`forecast weather request Error: ${err}`);
        return err;
    }
  };

  _onTextChange = (input) => {
    this.setState({
      textInput: input
    });
  };

  showModal = () => {
    this.setState({
      modalVisible: true
    });
  };

  hideModal = () => {
    this.setState({
      modalVisible: false
    });
  };

  getTextColour = (currentWeather) => {
    if(currentWeather === types.defaultInfo) {
      return types.whitesmoke;
    } else {
      return currentWeather.toLowerCase() === 'snow' ? types.grey : types.whitesmoke;
    }
  };

  render() {

    const { city, temperature, forecast, modalVisible, currentWeather, textColour} = this.state;

    return (
      <View style={styles.container}>

        <BackgroundImage
            currentWeather={currentWeather}
        />

        <CurrentWeatherPane
            city={city}
            currentWeather={currentWeather}
            temperature={temperature}
            textColour={textColour}
        />

        <ForecastList
            currentWeather={currentWeather}
            forecast={forecast}
            textColour={textColour}
        />

        <BottomSection
            getWeather={this.getWeather}
            onTextChange={this._onTextChange}
            showModal={this.showModal}
            closeModal={this.hideModal}
            modalVisible={modalVisible}
            currentWeather={currentWeather}
            textColour={textColour}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginBottom: 20,
    backgroundColor: '#F5FCFF',
  },

  topContainer: {
    marginTop: 50,
    margin: 10,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'lightskyblue',
    textAlign: 'center',
  },

  city: {
    height: 100,
    margin: 10,
  },
  weatherSummary: {
    height: 20,
    margin: 10,
  },
  temperature: {
    height: 100,
    margin: 10,
  },

  citySelect: {
    margin: 10,
    padding: 10,
    borderWidth: 1,
    borderColor: 'whitesmoke'
  },

  sectionContainer: {
    borderTopWidth: 0.5,
    borderColor: 'grey',
    marginLeft: 10,
    marginRight: 10
  }
});
