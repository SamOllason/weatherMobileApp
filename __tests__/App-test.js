/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../App';
import { shallow } from 'enzyme';
import * as types from '../types';
import fetchMock from 'fetch-mock';

// Only needed if using just Jest (as per top test)
import renderer from 'react-test-renderer';
types.defaultInfo,
// Using Jest
it('renders correctly, testing using just Jest', () => {
  renderer.create(<App />);
});

// Using Jest + Enzyme
describe('<App />', () => {
  it('renders correctly, testing using Jest+Enzyme', () => {
    expect.assertions(1);
    expect(shallow(<App/>)).toMatchSnapshot();
  });
});

// Using Jest + Enzyme
describe('Text input feature', () => {
  it('has a method to update textInput in component state', () => {
    expect.assertions(2);

    const wrapper = shallow(<App />);
    const instance = wrapper.instance();

    expect(wrapper.state('textInput')).toBe(types.defaultInfo);

    instance._onTextChange('theCity');
    expect(wrapper.state('textInput')).toBe('theCity');
  });
});

describe('Fetching Data', () => {

  it('should fetch CurrentWeather data from OpenWeatherMap', async () => {
    expect.assertions(3);

    // Setup
    const wrapper  = shallow(<App />);
    const instance = wrapper.instance();
    const url = `https://api.openweathermap.org/data/2.5/weather?q=${'London'}&APPID=${types.APIToken}&units=metric`;
    const fakeResponse = {
      name: 'London',
      main: {temp:11},
      weather: [
        { main: 'clear'},
        { main: 'sunny'},
      ]
    };

    // Use fetchMock here to mock the same response every time this url is called
    // This way we are implicitly mocking the fetch API.
    fetchMock.get(url, fakeResponse);

    const response = await instance.getCurrentWeather('London');

    // Check assertions
    expect(response.name).toEqual("London");
    expect(response.main.temp).toEqual(11);
    expect(response.weather[0].main).toEqual("clear");
  });

  it('should forecast data from OpenWeatherMap', async () => {
    expect.assertions(1);

    // Setup
    const wrapper  = shallow(<App />);
    const instance = wrapper.instance();
    const url = `https://api.openweathermap.org/data/2.5/forecast?q=${'London'}&APPID=${types.APIToken}&units=metric`

    const forecasts = [
      {
        "dt": 1534852800,
        "main": {
          "temp": 290.19,
          "temp_min": 290.19,
          "temp_max": 291.928,
          "pressure": 999.46,
          "sea_level": 1018.59,
          "grnd_level": 999.46,
          "humidity": 97,
          "temp_kf": -1.74
        },
        "weather": [
          {
            "id": 501,
            "main": "Rain",
            "description": "中雨",
            "icon": "10d"
          }
        ],
        "clouds": {
          "all": 92
        },
        "wind": {
          "speed": 5.01,
          "deg": 252.501
        },
        "rain": {
          "3h": 8.56
        },
        "sys": {
          "pod": "d"
        },
        "dt_txt": "2018-08-21 12:00:00"
      }
    ];

    const fakeResponse = {
      "cod": "200",
      "message": 0.0119,
      "cnt": 40,
      "list": forecasts
    };

    // Use fetchMock here to mock the same response every time this url is called
    // This way we are implicitly mocking the fetch API.
    fetchMock.get(url, fakeResponse);

    const response = await instance.getForecast('London');

    // Check assertions
    expect(response.list[0].dt_txt).toEqual("2018-08-21 12:00:00");
  });

  it('should recognise string code 200 is an error', () => {
    expect.assertions(3);

    const wrapper  = shallow(<App />);
    const instance = wrapper.instance();

    const fakeResponse1 = {
      cod: '200',
      message: 'an error occurred'
    };

    const isError1  = instance.checkDataStatusOk(fakeResponse1);
    expect(isError1).toBe(true);

    const fakeResponse2 = {
      cod: 200,
      message: 'an error occurred'
    };

    const isError2  = instance.checkDataStatusOk(fakeResponse2);
    expect(isError2).toBe(true);

    const fakeResponse3 = {
      cod: 400,
      message: '...'
    };

    const isError3  = instance.checkDataStatusOk(fakeResponse3);
    expect(isError3).toBe(false);

  });

  it('should update state with current weather data', () => {
    expect.assertions(6);

    const wrapper = shallow(<App />);
    const instance = wrapper.instance();
    expect(wrapper.state('city')).toBe(types.defaultInfo);
    expect(wrapper.state('temperature')).toBe(types.defaultInfo);
    expect(wrapper.state('currentWeather')).toBe(types.defaultInfo);

    const fakeResponse = {
      name: 'London',
      main: {temp:11},
      weather: [
          { main: 'clear'},
          { main: 'sunny'},
      ]
    };

    instance.updateCurrentWeather(fakeResponse);
    expect(wrapper.state('city')).toBe('London');
    expect(wrapper.state('temperature')).toBe(11);
    expect(wrapper.state('currentWeather')).toBe('clear');

  });

});

describe('Controlling InfoModal', () => {
  it('makes sure that InfoModal is not showing by default', () => {
    expect.assertions(1);

    const wrapper = shallow(<App />);
    expect(wrapper.state('modalVisible')).toBe(false);

  });

  it('has a method to show InfoModal', () => {
    expect.assertions(1);

    const wrapper  = shallow(<App />);
    const instance = wrapper.instance();

    instance.showModal();
    expect(wrapper.state('modalVisible')).toBe(true);
  });

  it('has a method to hide InfoModal', () => {
    expect.assertions(1);

    const wrapper  = shallow(<App />);
    const instance = wrapper.instance();

    instance.hideModal();
    expect(wrapper.state('modalVisible')).toBe(false);
  });
});

