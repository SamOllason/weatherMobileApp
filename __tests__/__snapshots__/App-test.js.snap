// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly, testing using Jest+Enzyme 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "marginBottom": 20,
    }
  }
>
  <BackgroundImage
    currentWeather="--"
  />
  <CurrentWeatherPane
    city="--"
    currentWeather="--"
    temperature="--"
  />
  <ForecastList
    currentWeather="--"
    forecast={Array []}
  />
  <BottomSection
    closeModal={[Function]}
    currentWeather="--"
    getWeather={[Function]}
    modalVisible={false}
    onTextChange={[Function]}
    showModal={[Function]}
  />
</View>
`;
